File: c++/c18_2_subtrees_and_paths.cpp
#include <bits/stdc++.h>
#include <gtest/gtest.h>
using namespace std;
using ll = long long;
using ull = unsigned long long;
#include "c09_2_binary_indexed_tree.h"
#include "c09_4_additional_techniques.h"

namespace {
	vector<ll> traversal(const vector<vector<ll>>& adj, ll x) {
		vector<ll> ret;
		auto dfs = [&](this auto self, ll s, ll e) -> void {
			ret.push_back(s);
			for (ll u : adj[s]) {
				if (u != e) {
					self(u, s);
				}
			}
			};
		dfs(x, -1);
		return ret;
	}

	struct SubtreeQueries {
		SubtreeQueries(const vector<vector<ll>>& adj, const vector<ll>& val, ll x)
			: pos(adj.size()), size(adj.size()), value(adj.size()) {
			ll cur = 0;
			auto& pos_ = pos;
			auto& size_ = size;
			auto dfs = [&](this auto self, ll s, ll e) -> ll {
				pos_[s] = cur++;
				ll sz = 1;
				for (ll u : adj[s]) {
					if (u != e) {
						sz += self(u, s);
					}
				}
				return size_[pos_[s]] = sz;
				};
			dfs(x, -1);
			for (ll i = 0; i < ssize(adj); ++i)
				value.add(pos[i] + 1, val[i]);
		}
		ll sum(ll k) const {
			ll p = pos[k];
			return value.sum(p + size[p]) - value.sum(p);
		}
		void set(ll k, ll x) {
			value.set(pos[k] + 1, x);
		}
	private:
		vector<ll> pos, size;
		BIT value;
	};

	struct PathQueries {
		PathQueries(const vector<vector<ll>>& adj, const vector<ll>& val, ll x)
			: pos(adj.size()), size(adj.size()), value(adj.size()) {
			ll cur = 0;
			auto& pos_ = pos;
			auto& size_ = size;
			auto& value_ = value;
			auto dfs = [&](this auto self, ll s, ll e, ll v) -> ll {
				pos_[s] = cur;
				v += val[s];
				value_.add(cur, cur, v);
				++cur;
				ll sz = 1;
				for (ll u : adj[s]) {
					if (u != e) {
						sz += self(u, s, v);
					}
				}
				return size_[pos_[s]] = sz;
				};
			dfs(x, -1, 0);
		}
		ll sum(ll k) const {
			return value[pos[k]];
		}
		void add(ll k, ll x) {
			ll p = pos[k];
			value.add(p, p + size[p], x);
		}
	private:
		vector<ll> pos, size;
		RangeUpdates value;
	};
}

TEST(C182SubtreesAndPaths, traversal) {
	EXPECT_EQ(traversal({ {} }, 0), vector<ll>{0});
	EXPECT_EQ(traversal({ {1,2,3,4},{0,5},{0},{0,6,7,8},{0},{1},{3},{3},{3} }, 0), (vector<ll>{0,1,5,2,3,6,7,8,4}));
}

TEST(C182SubtreesAndPaths, SubtreeQueries) {
	EXPECT_EQ(SubtreeQueries({ {} }, { 3 }, 0).sum(0), 3);
	SubtreeQueries sq({ {1,2,3,4},{0,5},{0},{0,6,7,8},{0},{1},{3},{3},{3} }, { 2,3,5,3,1,4,4,3,1 }, 0);
	EXPECT_EQ(sq.sum(3), 11);
	sq.set(6, 2);
	EXPECT_EQ(sq.sum(3), 9);
}

TEST(C182SubtreesAndPaths, PathQueries) {
	EXPECT_EQ(PathQueries({ {} }, { 3 }, 0).sum(0), 3);
	PathQueries pq({ {1,2,3,4},{0,5},{0},{0,6,7,8},{0},{1},{3},{3},{3} }, { 4,5,3,5,2,3,5,3,1 }, 0);
	EXPECT_EQ(pq.sum(6), 14);
	pq.add(3, 1);
	EXPECT_EQ(pq.sum(6), 15);
}

File: python/c18_2_subtrees_and_paths.py
import unittest
from c09_2_binary_indexed_tree import BIT
from c09_4_additional_techniques import RangeUpdates
import sys

sys.setrecursionlimit(max(sys.getrecursionlimit(),1<<20))

def traversal(adj,x):
    def dfs(s,e):
        yield s
        for u in adj[s]:
            if u!=e:
                yield from dfs(u,s)
    return dfs(x,0)

class SubtreeQueries:
    def __init__(self,adj,val,x):
        n=len(adj)
        cur=1
        self.pos=[None]*n
        self.size=[None]*n
        def dfs(s,e):
            nonlocal cur
            self.pos[s]=cur
            cur+=1
            size=1
            for u in adj[s]:
                if u!=e:
                    size+=dfs(u,s)
            self.size[self.pos[s]]=size
            return size
        dfs(x,0)
        self.value=BIT(n)
        for i in range(1,n):
            self.value.add(self.pos[i],val[i])
    def set(self,k,x):
        self.value[self.pos[k]]=x
    def sum(self,k):
        p=self.pos[k]
        return self.value.sum(p+self.size[p]-1)-self.value.sum(p-1)

class PathQueries:
    def __init__(self,adj,val,x):
        n=len(adj)
        cur=0
        self.pos=[None]*n
        self.size=[None]*(n-1)
        self.value=RangeUpdates(n-1)
        def dfs(s,e,v):
            nonlocal cur
            self.pos[s]=cur
            v+=val[s]
            self.value.add(cur,cur,v)
            cur+=1
            size=1
            for u in adj[s]:
                if u!=e:
                    size+=dfs(u,s,v)
            self.size[self.pos[s]]=size
            return size
        dfs(x,0,0)
    def add(self,k,x):
        p=self.pos[k]
        self.value.add(p,p+self.size[p]-1,x)
    def sum(self,k):
        return self.value[self.pos[k]]

class Test(unittest.TestCase):
    def test(self):
        adj=[[],[2,3,4,5],[1,6],[1],[1,7,8,9],[1],[2],[4],[4],[4]]
        self.assertEqual([*traversal([[],[]],1)],[1])
        self.assertEqual([*traversal(adj,1)],[1,2,6,3,4,7,8,9,5])

        self.assertEqual(SubtreeQueries([[],[]],[0,3],1).sum(1),3)
        sq=SubtreeQueries(adj,[0,2,3,5,3,1,4,4,3,1],1)
        self.assertEqual(sq.sum(4),11)
        sq.set(7,2)
        self.assertEqual(sq.sum(4),9)

        self.assertEqual(PathQueries([[],[]],[0,3],1).sum(1),3)
        pq=PathQueries(adj,[0,4,5,3,5,2,3,5,3,1],1)
        self.assertEqual(pq.sum(7),14)
        pq.add(4,1)
        self.assertEqual(pq.sum(7),15)
        self.assertEqual(pq.sum(5),6)

if __name__=='__main__':
    unittest.main()
